src/target/mspm0.c

```diff
@@ -0,0 +1,349 @@
+#define MSPM0_FLASHCTL_CMDTYPE_NOOP      0U
+#define MSPM0_FLASHCTL_CMDTYPE_PROG      1U
+#define MSPM0_FLASHCTL_CMDTYPE_ERASE     2U
+#define MSPM0_FLASHCTL_CMDTYPE_RDVERIFY  3U
+#define MSPM0_FLASHCTL_CMDTYPE_BLVERIFY  6U
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_1WORD  (0U << 4)
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_2WORDS (1U << 4)
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_4WORDS (2U << 4)
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_8WORDS (3U << 4)
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_SECTOR (4U << 4)
+#define MSPM0_FLASHCTL_CMDTYPE_SZ_BANK   (5U << 4)
+#define MSPM0_FLASHCTL_CMDEXEC_EXEC      1U
+#define MSPM0_FLASHCTL_STAT_DONE         0x01U
+#define MSPM0_FLASHCTL_STAT_CMDPASS      0x02U
+
+typedef struct mspm0_flash {
+	target_flash_s target_flash;
+	uint32_t banks;
+} mspm0_flash_s;
+
+static bool mspm0_flash_erase(target_flash_s *flash, target_addr_t addr, size_t length);
+static bool mspm0_flash_write(target_flash_s *flash, target_addr_t dest, const void *src, size_t length);
+static bool mspm0_mass_erase(target_s *target);
+
+#if MSPM0_CONFIG_FLASH_DUMP_SUPPORT
+static bool mspm0_dump_factory_config(target_s *const target, const int argc, const char **const argv);
+static bool mspm0_dump_bcr_config(target_s *const target, const int argc, const char **const argv);
+
+static command_s mspm0_cmds_list[] = {{"dump_factory", mspm0_dump_factory_config, "Display FACTORY registers"},
+	{"dump_bcr", mspm0_dump_bcr_config, "Display NONMAIN (BCR/BSL) registers"},
+	/* {"write_nm", mspm0_write_non_main, "Write data to NONMAIN (configuration) flash. write_nm <address> <value>"}, */
+	{NULL, NULL, NULL}};
+
+typedef struct conf_register {
+	uint16_t offset : 12;
+	uint16_t words : 4;
+	char const *id;
+} conf_register_s;
+
+static conf_register_s mspm0_factory_regs[] = {{0x00, 1, "TRACEID"}, {0x04, 1, "DEVICEID"}, {0x08, 1, "USERID"},
+	{0x0c, 1, "BSLPIN_UART"}, {0x10, 1, "BSLPIN_I2C"}, {0x14, 1, "BSLPIN_INVOKE"}, {0x18, 1, "SRAMFLASH"},
+	{0x3c, 1, "TEMP_SENSE0"}, {0x7c, 1, "BOOTCRC"}, {0, 0, NULL}};
+
+static conf_register_s mspm0_bcr_regs[] = {{0x00, 1, "BCRCONFIGID"}, {0x04, 1, "BOOTCFG0"}, {0x08, 1, "BOOTCFG1"},
+	{0x0c, 4, "PWDDEBUGLOCK"}, {0x1c, 4, "BOOTCFG2"}, {0x20, 1, "BOOTCFG3"}, {0x24, 4, "PWDMASSERASE"},
+	{0x34, 4, "PWDFACTORYRESET"}, {0x44, 1, "FLASHSWP0"}, {0x48, 1, "FLASHSWP1"}, {0x4c, 1, "BOOTCFG4"},
+	{0x50, 1, "APPCRCSTART"}, {0x54, 1, "APPCRCLENGTH"}, {0x58, 1, "APPCRC"}, {0x5c, 1, "BOOTCRC"},
+	{0x100, 1, "BSLCONFIGID"}, {0x104, 1, "BSLPINCFG0"}, {0x108, 1, "BSLPINCFG1"}, {0x10c, 1, "BSLCONFIG0"},
+	{0x110, 8, "BSLPW"}, {0x130, 1, "BSLPLUGINCFG"}, {0x134, 4, "BSLPLUGINHOOK"}, {0x144, 1, "PATCHHOOKID"},
+	{0x148, 1, "SBLADDRESS"}, {0x14c, 1, "BSLAPPVER"}, {0x150, 1, "BSLCONFIG1"}, {0x154, 1, "BSLCRC"}, {0, 0, NULL}};
+
+static void mspm0_dump_regs(target_s *const target, const conf_register_s *regs, uint32_t base)
+{
+	for (conf_register_s const *r = regs; r->id; ++r) {
+		tc_printf(target, "%15s: ", r->id);
+		for (int q = 0; q < r->words; ++q) {
```

Comment:
Do not use `int` for this loop. This performs a signed-unsigned conversion and opens the code up to bugs we do not want to tango with here. Either use `uint16_t` to match the intended underlying type of `words`, or `size_t`.